File: android/app/src/main/kotlin/hid/DeviceModel.kt
package hid

data class DeviceModel(
    val id: String,
    val productName: String,
    val productIdMM: Int,
    val legacyUsbProductId: Int
)

object Devices {
    private val devicesList = listOf(
        DeviceModel("blue", "Ledger Blue", 0x00, 0x0000),
        DeviceModel("nanoS", "Ledger Nano S", 0x10, 0x0001),
        DeviceModel("nanoX", "Ledger Nano X", 0x40, 0x0004),
        DeviceModel("nanoSP", "Ledger Nano S Plus", 0x50, 0x0005),
        DeviceModel("stax", "Ledger Stax", 0x60, 0x0006),
        DeviceModel("europa", "Ledger Flex", 0x70, 0x0007),
        DeviceModel("apex", "Ledger Apex", 0x80, 0x0008)
    )

    fun identifyUSBProductId(usbProductId: Int): DeviceModel? {
        val legacy = devicesList.find { it.legacyUsbProductId == usbProductId }
        if (legacy != null) return legacy
        val mm = usbProductId shr 8
        return devicesList.find { it.productIdMM == mm }
    }
}

File: lib/ledger/models/device_model.dart
import 'package:collection/collection.dart';

class BleSpec {
  final String serviceUuid;
  final String notifyUuid;
  final String writeUuid;
  final String? writeCmdUuid;

  const BleSpec({
    required this.serviceUuid,
    required this.notifyUuid,
    required this.writeUuid,
    this.writeCmdUuid,
  });
}

class DeviceModel {
  final String id;
  final String productName;
  final int productIdMM;
  final int legacyUsbProductId;
  final List<BleSpec>? bluetoothSpec;

  const DeviceModel({
    required this.id,
    required this.productName,
    required this.productIdMM,
    required this.legacyUsbProductId,
    this.bluetoothSpec,
  });
}

class Devices {
  static const List<DeviceModel> _devicesList = [
    DeviceModel(
      id: 'blue',
      productName: 'Ledger Blue',
      productIdMM: 0x00,
      legacyUsbProductId: 0x0000,
    ),
    DeviceModel(
      id: 'nanoS',
      productName: 'Ledger Nano S',
      productIdMM: 0x10,
      legacyUsbProductId: 0x0001,
    ),
    DeviceModel(
      id: 'nanoX',
      productName: 'Ledger Nano X',
      productIdMM: 0x40,
      legacyUsbProductId: 0x0004,
      bluetoothSpec: [
        BleSpec(
          serviceUuid: "13d63400-2c97-0004-0000-4c6564676572",
          notifyUuid: "13d63400-2c97-0004-0001-4c6564676572",
          writeUuid: "13d63400-2c97-0004-0002-4c6564676572",
          writeCmdUuid: "13d63400-2c97-0004-0003-4c6564676572",
        ),
      ],
    ),
    DeviceModel(
      id: 'nanoSP',
      productName: 'Ledger Nano S Plus',
      productIdMM: 0x50,
      legacyUsbProductId: 0x0005,
    ),
    DeviceModel(
      id: 'stax',
      productName: 'Ledger Stax',
      productIdMM: 0x60,
      legacyUsbProductId: 0x0006,
      bluetoothSpec: [
        BleSpec(
          serviceUuid: "13d63400-2c97-6004-0000-4c6564676572",
          notifyUuid: "13d63400-2c97-6004-0001-4c6564676572",
          writeUuid: "13d63400-2c97-6004-0002-4c6564676572",
          writeCmdUuid: "13d63400-2c97-6004-0003-4c6564676572",
        ),
      ],
    ),
  ];

  static DeviceModel? identifyUSBProductId(int usbProductId) {
    try {
      return _devicesList
          .firstWhere((d) => d.legacyUsbProductId == usbProductId);
    } catch (_) {
      final mm = usbProductId >> 8;
      try {
        return _devicesList.firstWhere((d) => d.productIdMM == mm);
      } catch (_) {
        return null;
      }
    }
  }

  static DeviceModel? identifyBluetoothServiceUuid(String uuid) {
    final lowerUuid = uuid.toLowerCase();
    return _devicesList.firstWhereOrNull((d) =>
        d.bluetoothSpec
            ?.any((spec) => spec.serviceUuid.toLowerCase() == lowerUuid) ??
        false);
  }
}
